opacus/privacy_engine.py

```diff
@@ -203,28 +212,33 @@ def _prepare_model(
         *,
         batch_first: bool = True,
         loss_reduction: str = "mean",
-    ) -> GradSampleModule:
+        grad_sample_mode: str = "hooks",
```

Comment:
How do you feel about making subclasses of PrivacyEngine and attaching it to one kind of GSM? Say `PrivacyEngine` and `PrivacyEngineWithExpandedWeights`?

Pros: (1) avoids pesky `if-else` (2) makes it easier and clearer to the end user and developer to follow since one class only has to deal with one implementation (3) extensible: we can have `PrivacyEngineWithFuncTorch` in the future if needed without modifying the existing code.

Cons: (1) More user facing classes (2) unnecessary if we want to eventually deprecate one of the GSM approaches. (3) <I'm sure I'm missing some, but can't think of any; please comment with more if you see it>

Reply:
To be honest, I'm not a big fan.

At this point the only customisation required to switch between EW and hooks is the class name, and I think PrivacyEngine already acts like sort of factory, mapping parameters (e.g. `distributed` or `clipping`) to the most appropriate implementation.

>  avoids pesky if-else

True, but I don't see it as a big nuisance

> makes it easier and clearer to the end user and developer to follow since one class only has to deal with one implementation

I would argue, that controlling behaviour with a string-argument is an accepted practice, and for most users `grad_sample_mode` is an insignificant implementation detail. And if you face an issue with the default implementation, would it be easier for you to pass an extra argument or switch to a different class?

> extensible: we can have PrivacyEngineWithFuncTorch in the future if needed without modifying the existing code.

I don't think we have good enough picture on how true-functorch opacus would look like. I agree it's a possible future, but I'm struggling to take this into account due to amount of ambiguity

 > More user facing classes

++

> unnecessary if we want to eventually deprecate one of the GSM approaches.

It's a good question whether we will. Ideally, we'd want to, but it does feel like we'd have to live with both for a while

>  <I'm sure I'm missing some, but can't think of any; please comment with more if you see it>

For me the strongest point is that with `grad_sample_mode` as an argument we have better control over default behaviour. For example at some point we'll be able to switch ExpandedWeights by default when they reach certain point of maturity

Reply:
I'm not entirely convinced, but I think you have said enough to make me not block you :)

My annoyance is largely due to this:
> I would argue, that controlling behaviour with a string-argument is an accepted practice, and for most users grad_sample_mode is an insignificant implementation detail. And if you face an issue with the default implementation, would it be easier for you to pass an extra argument or switch to a different class?

1. I don't think this is an insignificant implementation detail. If the difference were just one being faster than the other, I would have agreed, but here, the two implementations differ in what they support too. This, to me, is a huge change.
2.  I typically think of flags as changing a behaviour of an execution flow as it is flowing, not changing the whole component. Eg, accountants in opacus right now are different classes, and rightly so. It could have easily been an `if-else` gated logic enabled by a `string` flag, but that wouldn't have been clean.


However, as I said above, my annoyance is not terrible enough to block the PR.

Reply:
I would gladly continue the discussion and not rush the PR until we find a common ground.

> Eg, accountants in opacus right now are different classes, and rightly so.

Buuut, accountants are controlled with a string argument (see `PrivacyEngine` constructor). They are implemented as different classes, but exposed to the client through factory-like behaviour in `PrivacyEngine`

I agree that `GradSampleModule` implementation is important and does potentially change the behaviour. Perhaps my argument on it being merely an implementation detail doesn't hold water and I withdraw it.

However, this doesn't change my position on subclassing PrivacyEngine. If we go back to 1.0 designs, one of the pain points for us was the combinatorial explosion from having multiple optimizers and accountants. If we were to use `PrivacyEngine` subclasses, we'd need to create at least 16 of them: 2 accountants, 2 clipping strategies, disrtibuted/non-distributed, poisson/non-poisson (and now different grad samplers).

We've played with it in PR #254, but it ended up being too cumbersome: navigating the code became too hard.

And I don't think it make sense to have new `PrivacyEngine` subclasses for different GradSampleModules, but not optimizers or data loaders

Reply:
> Buuut, accountants are controlled with a string argument (see PrivacyEngine constructor). They are implemented as different classes, but exposed to the client through factory-like behaviour in PrivacyEngine

Touché. A small nit in that case, accept the gsm_mode as an arg to `PrivacyEngine` rather than `make_private`. To follow the same pattern?